pyro/infer/abstract_infer.py

```diff
@@ -285,23 +286,57 @@ class TracePredictive(TracePosterior):
     :param TracePosterior posterior: trace posterior instance holding
         samples from the model's approximate posterior.
     :param int num_samples: number of samples to generate.
+    :param list(str) resample_sites: sites to resample from in posterior predictive distribution.
+                                     By default all sites are used.
     """
-    def __init__(self, model, posterior, num_samples):
+    def __init__(self, model, posterior, num_samples, resample_sites=None):
         self.model = model
         self.posterior = posterior
         self.num_samples = num_samples
+        self.resample_sites = resample_sites
         super(TracePredictive, self).__init__()
 
     def _traces(self, *args, **kwargs):
```

Comment:
I don't think we should need the `resample_posterior` poutine at all? We can probably simplify this method to:

```python
        if not self.posterior.exec_traces:
            self.posterior.run(*args, **kwargs)
        data_trace = poutine.trace(self.model).get_trace(*args, **kwargs)
        for _ in range(self.num_samples):
            model_trace = self.posterior()
            self._adjust_to_data(model_trace, data_trace)
            replayed_trace = poutine.trace(poutine.replay(self.model, model_trace)).get_trace(*args, **kwargs)
            yield (replayed_trace, 0., 0)
```

Reply:
OK, I will remove the `resample_posterior` poutine.